receiver/bigipreceiver/client.go

```diff
@@ -0,0 +1,321 @@
+	}
+
+	return &bigipClient{
+		client:       httpClient,
+		hostEndpoint: cfg.Endpoint,
+		creds: bigipCredentials{
+			username: cfg.Username,
+			password: cfg.Password,
+		},
+		logger: logger,
+	}, nil
+}
+
+// HasToken checks to see if an auth token has been set for the client
+func (c *bigipClient) HasToken() bool {
+	return c.token != ""
+}
+
+// GetNewToken makes an appropriate call to the iControl REST login endpoint and sets the returned token on the bigipClient
+func (c *bigipClient) GetNewToken(ctx context.Context) error {
+	var tokenDetails *models.TokenDetails
+
+	if err := c.post(ctx, loginPath, &tokenDetails); err != nil {
+		c.logger.Debug("Failed to retrieve api token", zap.Error(err))
+		return err
+	}
+
+	c.token = tokenDetails.Token.Token
+	return nil
+}
+
+// GetVirtualServers makes calls to both the standard and statistics version of the virtual servers endpoint.
+// It combines this info into one object and returns it.
+func (c *bigipClient) GetVirtualServers(ctx context.Context) (*models.VirtualServers, error) {
+	// get standard Virtual Server details
+	var virtualServers *models.VirtualServers
+
+	if err := c.get(ctx, virtualServersStatsPath, &virtualServers); err != nil {
+		c.logger.Debug("Failed to retrieve virtual servers", zap.Error(err))
+		return nil, err
+	}
+
+	// get statistic virtual server details and combine them
+	var virtualServersDetails *models.VirtualServersDetails
+
+	if err := c.get(ctx, virtualServersPath, &virtualServersDetails); err != nil {
+		c.logger.Warn("Failed to retrieve virtual servers properties", zap.Error(err))
+		return virtualServers, nil
+	}
+
+	return addVirtualServerPoolDetails(virtualServers, virtualServersDetails), nil
+}
+
+// GetPools makes a call the statistics version of the pools endpoint and returns the data.
+func (c *bigipClient) GetPools(ctx context.Context) (*models.Pools, error) {
+	var pools *models.Pools
+
+	if err := c.get(ctx, poolsStatsPath, &pools); err != nil {
+		c.logger.Debug("Failed to retrieve pools", zap.Error(err))
+		return nil, err
+	}
+
+	return pools, nil
+}
+
+// GetPoolMembers takes in a list of all Pool data. It then iterates over this list to make a call to the statistics version
+// of each pool's pool members endpoint. It accumulates all of this data into a single pool members object and returns it.
+func (c *bigipClient) GetPoolMembers(ctx context.Context, pools *models.Pools) (*models.PoolMembers, error) {
+	var (
+		poolMembers         *models.PoolMembers
+		combinedPoolMembers *models.PoolMembers
+	)
+	errorReceived := false
+	collectedPoolMembers := false
+
+	// for each pool get pool member info and aggregate it into a single spot
+	for poolURL := range pools.Entries {
+		poolMemberPath := strings.TrimPrefix(poolURL, "https://localhost")
+		poolMemberPath = strings.TrimSuffix(poolMemberPath, "/stats") + poolMembersStatsPathSuffix
+
+		if err := c.get(ctx, poolMemberPath, &poolMembers); err != nil {
```

Comment:
These errors should probably be captured in a multierr and returned, for the above caller to add to a scraper errors as a partial error.